### PR TITLE
Fix cluster event bus

### PR DESCRIPTION
**Pull request by joschi**
Documents in capped collections cannot be updated if their size would change by the update, which is exactly what happened with the "consumers" array.

This fixes the original change by creating a non-capped collection and cleaning up old events after 1 hour. We might make this configurable in the future, but for now it should be fine. The "cluster_events" collection is now being polled instead of tailed with a tailable cursor (which only works in combination with a capped collection).

### CODE REVIEW

**Comment by bernd** [comment]
I personally would rather use a regular collection and a retention job to clean it up than using timestamps for this. We already have that for our node status and it is not really nice. There are lots of things that can go wrong when using timestamps. (i.e. unsynced clocks, gc pauses)

**Comment by kroepke** [comment]
I agree that a non-capped collection and cleaning job would be more robust than working with timestamps.
Clock skew/unsynced clocks pose a serious problem with using timestamps, as Bernd noted this has happened before.

The cleanup should be simple by having an index on the date and be limited by count as well. That way a node can also process events that occurred during its own downtime, whether an event has an actual effect then only depends on the consumer but not on the lookup mechanism from mongodb.

**Comment by joschi** [comment]
@kroepke @bernd Having played around with it, I tend to agree with you. The problem with this is, that [tailable cursors](http://docs.mongodb.org/manual/tutorial/create-tailable-cursor/) can only be used with capped collections in MongoDB.

This means that we would have to resort to stupid polling of the `cluster_events` collection.

**Comment by bernd** [comment]
I guess we have to fall back to polling then. :disappointed:

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java [hunk 1]
```diff
@@ -61,70 +55,126 @@
...
-                coll = db.createCollection(COLLECTION_NAME, options);
-            } catch (MongoException e) {
-                // Error code 48 == "collection already exists"
-                if(e.getCode() == 48) {
-                    coll = db.getCollection(COLLECTION_NAME);
+        DBCollection coll = db.getCollection(COLLECTION_NAME);
+
+        coll.createIndex(DBSort.desc("timestamp"));
+        coll.createIndex(DBSort.asc("producer"));
+        coll.createIndex(DBSort.asc("consumers"));
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: Does this work pre-mongodb 2.6?

The query below is a and-query with a sort on timestamp. I think this should be a single index (timestamp desc, producers asc, consumers asc) instead.

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java [hunk 2]
```diff
@@ -138,14 +188,10 @@ static DBCollection prepareCollection(final MongoConnection mongoConnection) {
         and.add(consumersClause);
         final DBObject query = new BasicDBObject("$and", and);
 
-        return dbCollection.find(query)
-                .sort(DBSort.asc("$natural"))
-                .addOption(Bytes.QUERYOPTION_TAILABLE)
-                .addOption(Bytes.QUERYOPTION_AWAITDATA);
+        return dbCollection.find(query).sort(DBSort.desc("timestamp"));
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: why desc?
Shouldn't events be processed in the time order they were created?

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventCleanupPeriodical.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    protected Logger getLogger() {
+        return LOG;
+    }
+
+    @Override
+    public void doRun() {
+        try {
+            LOG.debug("Removing stale events from MongoDB collection \"{}\"", COLLECTION_NAME);
+
+            final long now = DateTime.now(DateTimeZone.UTC).getMillis() - maxEventAge;
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventCleanupPeriodical.java]: (extremely) minor nitpick: the content of now isn't now :)

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java (hunk 2 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: Yes, indeed. :blush:

**Comment by kroepke** [comment]
lgtm!

**Pull request merged**